homeassistant/components/godice/dice.py

```diff
@@ -0,0 +1,151 @@
+        """Subscribe for receiving notifications with new numbers when Dice is rolled."""
+        await self._dice.subscribe_number_notification(callback)
+
+    async def pulse_led(self, *args, **kwargs):
+        """Pulse built-in LEDs."""
+        await self._dice.pulse_led(*args, **kwargs)
+
+
+class NoopDice:
+    """No-op Dice impl used when GoDice is disconnected."""
+
+    async def connect(self, disconnect_cb):
+        """Connect."""
+
+    async def disconnect(self):
+        """Disconnect."""
+
+    async def get_color(self):
+        """Get Dice color."""
+
+    async def get_battery_level(self):
+        """Get Dice battery level."""
+
+    async def subscribe_number_notification(self, _callback):
+        """Subscribe for receiving notifications with new numbers when Dice is rolled."""
+
+    async def pulse_led(self, *args, **kwargs):
+        """Pulse built-in LEDs."""
+
+
+class DiceProxy(DiceDelegate):
+    """Proxy between HA and Dice, manages connection establishment."""
+
+    def __init__(self, hass: HomeAssistant, entry: ConfigEntry) -> None:
+        """Init connection data."""
+        self._mac = entry.data["address"]
+        self._hass = hass
+        self._dice = NoopDice()
+        self._conn_lock = asyncio.Lock()
+        self._bledev: BLEDevice | None = None
+        self._bledev_upd_cancel = bluetooth.async_register_callback(
+            hass,
+            self._upd_bledev,
+            bluetooth.BluetoothCallbackMatcher({bluetooth.match.ADDRESS: self._mac}),
+            bluetooth.BluetoothScanningMode.ACTIVE,
+        )
+        self._disconnect_cb = None
+        self._client = None
+        self._disconnect_handler = self._on_connection_lost_while_connecting_handler
+
+    async def connect(self, disconnect_cb):
+        """Connect to the Dice."""
+        async with self._conn_lock:
+            _LOGGER.debug("Connecting")
+            self._disconnect_cb = disconnect_cb
+            self._bledev = self._bledev or bluetooth.async_ble_device_from_address(
+                self._hass, self._mac
+            )
+            try:
+                self._client = await establish_connection(
+                    client_class=BleakClient,
+                    device=self._bledev,
+                    name=self._mac,
+                    disconnected_callback=self._on_disconnected_handler,
+                    max_attempts=3,
+                    use_services_cache=True,
+                )
+                dice = godice.create(self._client, godice.Shell.D6)
+                await dice.connect()
+                self._dice = dice
+                self._disconnect_handler = (
+                    self._on_connection_lost_after_connected_handler
+                )
+                _LOGGER.debug("Connection completed")
+            except BleakError as e:
+                _LOGGER.debug("Connection attempts timed out")
+                await close_stale_connections(self._bledev)
+                raise e
+
```

Comment:
Please move all the connection logic into the library.

We want as little code as possible in Home Assistant that communicates with the device

https://developers.home-assistant.io/docs/api_lib_index?_highlight=pypi#trying-your-library-inside-home-assistant

Reply:
thanks for your help! Addressed, please check